src/ansys/dpf/composites/solid_stack_results.py

```diff
@@ -0,0 +1,224 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Functions to retrieve through-the-thickness results of a stack of solid elements."""
+from collections.abc import Sequence
+
+import ansys.dpf.core as dpf
+import numpy as np
+
+from ansys.dpf.composites.constants import Spot, component_index_from_name
+from ansys.dpf.composites.failure_criteria import FailureModeEnum
+from ansys.dpf.composites.layup_info import ElementInfo, ElementInfoProvider, SolidStack
+
+from .sampling_point_types import FailureResult
+from .select_indices import get_selected_indices
+
+__all__ = (
+    "get_through_the_thickness_failure_results",
+    "get_through_the_thickness_results",
+)
+
+MAX_RESERVE_FACTOR = 1000.0
+SOLID_SPOTS = [Spot.BOTTOM, Spot.TOP]
+
+
+def _irf2rf(irf: float) -> float:
+    if irf == 0.0:
+        return MAX_RESERVE_FACTOR
+
+    rf = 1.0 / irf
+    return min(rf, MAX_RESERVE_FACTOR)
+
+
+def _irf2mos(irf: float) -> float:
+    rf = _irf2rf(irf)
+    return rf - 1.0
+
+
+def _get_element_info(element_info_provider: ElementInfoProvider, element_id: int) -> ElementInfo:
+    e_info = element_info_provider.get_element_info(element_id)
+    if not e_info:
+        raise RuntimeError(f"Could not retrieve element info for {element_id}")
+    return e_info
+
+
+def get_through_the_thickness_failure_results(
+    solid_stack: SolidStack,
+    element_info_provider: ElementInfoProvider,
+    irf_field: dpf.Field,
+    failure_mode_field: dpf.Field,
+) -> list[FailureResult]:
+    """
+    Get through-the-thickness failure results of the solid stack.
+
+    The maximum IRF is extracted at the bottom and top of each ply for each element in
+    the stack. The result contains the failure mode, IRF, reserve factor and margin of
+    safety for each spot and ply. The inputs irf_field and failure_mode_field are
+    expected to have location element_nodal (no data reduction) with data at the
+    bottom and top of each layer (ply).
+
+    In case of drop-off or cut-off elements, which are non-layered, the maximum IRF is extracted.
+
+    Example:
```

Comment:
In the numpy docstring style, the examples should use a heading with underline: https://numpydoc.readthedocs.io/en/latest/format.html#examples